File: heap.py
import math
import sys
from time import time
import random
def heapsort(a, count):
	heapify(a, count)
	end = count - 1
	while (end > 0):
		swap(a,end,0)
		end = end - 1
		siftdown(a, 0, end)
def heapify(a, count):
	start = parent(count - 1)
	while (start >= 0):
		siftdown(a, start, count - 1)
		start = start - 1
def siftdown(a, start, end):
		root = start
		while (left(root) <= end):
			child = left(root)
			_swp = root
			if a[_swp] < a[child]:
				_swp = child
			if (child + 1 <= end and a[_swp] < a[child + 1]):
				_swp = child + 1
			if _swp == root:
				return	
			else:
				swap(a,root,_swp)
				root = _swp
def swap(a,i,j):
	a[i],a[j] = a[j],a[i]
	return
def siftup(a, start, end):
	child = end
	while child > start:
		pa = parent(child)
		if a[pa] < a[child]:
			swap(a,pa,child)
			child = pa
		else:
			return
def parent(i):
	return int(math.floor((i-1)/2))
def left(i):
	return 2*i + 1
def right(i):
	return 2*i + 2
start = time()
a=[]
for i in range(int(sys.argv[1])):
	a.append(i)

random.shuffle(a)
heapsort(a,int(sys.argv[1]))
print a
end = time()
print ("this took " + str(end - start))

File: heap.cpp
#include <iostream>
#include <omp.h>
#include <random>
#include <algorithm>
#include <string>
#include <vector>
#include <chrono>
#include <stdlib.h>
#include <math.h>

int parent(int i) { 
	return (int) floor( ((float)(i-1)) / 2);
}

int left(int i) {
	return 2*i + 1;
}

int right(int i) {
	return 2*i + 2;
}

void siftdown(std::vector<int> &a, int start, int end) {
	int root = start;
	while (left(root) <= end) {
		int child = left(root);
		int swap = root;
		if (a[swap] < a[child]) {
			swap = child;
		}
		if (child + 1 <= end && a[swap] < a[child + 1]) {
			swap = child + 1;
		}
		if (swap == root) {
			return;
		}
		else {
			std::swap(a[root],a[swap]);
			root = swap;
		}
	}
}

void heapify(std::vector<int> &a, int count) {
	int start = parent(count - 1);
	while (start >= 0) {
		siftdown(a,start,count - 1);
		start = start - 1;
	}
}

void heapsort(std::vector<int> &a, int count) {
	heapify(a,count);
	int end = count - 1;
	while (end > 0) {
		std::swap(a[end],a[0]);
		end = end - 1;
		siftdown(a, 0, end);
	}
}

void shuffle_heap(std::vector<int> &s) {
	std::random_device rd;
	auto seed = std::chrono::system_clock::now().time_since_epoch().count();
	auto engine = std::default_random_engine(seed);
	std::shuffle(std::begin(s), std::end(s), engine);
}

int main (int argc, char ** argv){
	std::vector<int> heap;
	unsigned int c = atoi(argv[1]);
	for (unsigned int i = 0; i < c; i++) {
		heap.push_back(i);
	}
	shuffle_heap(heap);
	double start = omp_get_wtime();
	heapsort(heap,c);
	//for (auto a: heap) {
//		std::cout << a << "\n"; //<< std::endl;
//	}
	double end = omp_get_wtime();
	double diff = end-start;
	std::cout << "This took " << diff << " seconds to execute!";
}
